aptos-move/aptos-aggregator/src/bounded_math.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright © Aptos Foundation
+
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::vm_status::StatusCode;
+use aptos_logger::error;
+
+use crate::delta_change_set::{EEXPECTED_OVERFLOW, EEXPECTED_UNDERFLOW, ECODE_INVARIANT_BROKEN};
+
+
+
+#[derive(Debug)]
+pub enum BoundedMathError {
+    Overflow,
+    Underflow,
+}
+
+/// Error for delta application. Can be used by delta partial functions
+/// to return descriptive error messages and an appropriate error code.
+pub(crate) fn abort_error(message: impl ToString, code: u64) -> PartialVMError {
+    PartialVMError::new(StatusCode::ABORTED)
+        .with_message(message.to_string())
+        .with_sub_status(code)
+}
+
+impl From<BoundedMathError> for PartialVMError {
+    fn from(err: BoundedMathError) -> Self {
+        match err {
+            BoundedMathError::Overflow => abort_error(
+                "Overflow",
+                EEXPECTED_OVERFLOW,
+            ),
+            BoundedMathError::Underflow => abort_error(
+                "Underflow",
+                EEXPECTED_UNDERFLOW,
+            ),
+        }
+    }
+}
+
+pub type BoundedMathResult<T> = ::std::result::Result<T, BoundedMathError>;
+
+pub fn ok_overflow<T>(value: BoundedMathResult<T>) -> BoundedMathResult<Option<T>> {
+    match value {
+        Ok(value) => Ok(Some(value)),
+        Err(BoundedMathError::Overflow) => Ok(None),
+        Err(BoundedMathError::Underflow) => Err(BoundedMathError::Underflow),
+    }
+}
+
+pub fn ok_underflow<T>(value: BoundedMathResult<T>) -> BoundedMathResult<Option<T>> {
+    match value {
+        Ok(value) => Ok(Some(value)),
+        Err(BoundedMathError::Overflow) => Err(BoundedMathError::Overflow),
+        Err(BoundedMathError::Underflow) => Ok(None),
+    }
+}
+
+pub fn expect_ok<T>(value: BoundedMathResult<T>) -> PartialVMResult<T> {
```

Comment:
We are mapping even overflow and underflow errors to ECODE_INVARIANT_BROKEN? overflow and underflow errors are not bugs in the code as specified in the below error message right?

Reply:
we are mapping them when expect_ok is explicitly called. 

so the caller decides if it code invariant, or regular case. 

for example:
```
                    let overflow_value = expect_ok(ok_overflow(addition_deltavalue(
                        input,
                        delta,
                        self.max_value,
                    )))?;
```

this will make overflow_value be None on Overflow (due to ok_overflow call), but raise ECODE_INVARIANT_BROKEN on underflow (due to expect_ok call around it)